doc/TPID/TPID_SAI_proposal.md

```diff
@@ -0,0 +1,272 @@
+and 0xAAB8) that the user specified. Although ideally any TPID values
+can be used but to simplify the management of TPID resource tracking by
+SONiC, SAI can claim TPID attribute is supported when it allows setting
+of any of the above commonly used TPID values on any given port/LAG
+interface. SAI implementation can optionally support more TPID values
+beyond these 4 commonly used TPIDs for VLAN Tagging purpose.
+
+The dynamic setting of the TPID value on a port/LAG shall use the
+existing port/LAG set SAI attribute API with the new TPID port/LAG
+attribute specified.
+
+The TPID attribute value shall use the following
+sai\_attribute\_value\_t:
+```
+/** @validonly meta->attrvaluetype == SAI_ATTR_VALUE_TYPE_UINT16 */
+sai_uint16_t u16;
+```
+
+Setting Port TPID shall call the following existing SAI API with
+sai\_attr\_id\_t set to SAI\_PORT\_ATTR\_TPID:
+```
+/**
+ * @brief Set port attribute value.
+ *
+ * @param[in] port_id Port id
+ * @param[in] attr Attribute
+ *
+ * @return #SAI_STATUS_SUCCESS on success, failure status code on error
+ */
+typedef sai_status_t (*sai_set_port_attribute_fn)(
+        _In_ sai_object_id_t port_id,
+        _In_ const sai_attribute_t *attr);
+```
+
+Setting LAG TPID shall call the following existing SAI API with
+sai\_attr\_id\_t set to SAI\_LAG\_ATTR\_TPID:
+```
+/**
+ * @brief Set LAG Attribute
+ *
+ * @param[in] lag_id LAG id
+ * @param[in] attr Structure containing ID and value to be set
+ *
+ * @return #SAI_STATUS_SUCCESS on success, failure status code on error
+ */
+typedef sai_status_t (*sai_set_lag_attribute_fn)(
+        _In_ sai_object_id_t lag_id,
+        _In_ const sai_attribute_t *attr);
+```
+
+**SAI TPID HW Resource Capability Query Support:**
+--------------------------------------------------
+
+For some hardware SKU there may be a HW limitation as to TPID
+configuration on the physical interface is supported. This limitation
+shall be made known to the SONiC application so that a configuration
+precheck can be performed as part of any TPID config validation. The new
+switch attribute "SAI\_SWITCH\_ATTR\_TPID\_CONFIGURABLE" allows the
+
+HW SKU to specify whether it support this feature to set TPID to any of
+the commonly used TPID values to any port/LAG specified. At boot up
+time, this attribute is queried by SONIC Application. If SAI supports
+this TPID config attribute, the response value for this new attribute
+should be set to "1". If TPID config is NOT supported, the corresponding
+api return status should fail with SAI_STATUS_ATTR_NOT_SUPPORTED_xxx or 
+SAI_STATUS_ATTR_NOT_IMPLEMENTED_xxx. 
+
+```
+/**
+ * @brief Attribute Id in sai_set_switch_attribute() and
+ * sai_get_switch_attribute() calls
+ */
+typedef enum _sai_switch_attr_t
+{
+    …
+    /**
+     * @brief Up to 4 TPIDs can be configured on this switch to any port/LAG
+     *
+     * @type bool
+     * @flags READ_ONLY
+     */
+    SAI_SWITCH_ATTR_TPID_CONFIGURABLE,

```

Comment:
There is already general attribute capability query, sai_query_attribute_capability()
why not use the general mechanism instead if tpid_configurable ?

Reply:
Are you referring to 
sai_status_t sai_query_attribute_capability(_In_ sai_object_id_t         switch_id,
                                            _In_ sai_object_type_t       object_type,
                                            _In_ sai_attr_id_t           attr_id,
                                            _Out_ sai_attr_capability_t *attr_capability)
just want to clarify my understanding of this API.  So for my purpose you are suggesting to  call this API at boot up time  and pass the switch_id, the object_type set to "SAI_OBJECT_TYPE_PORT" or "SAI_OBJECT_TYPE_LAG", and the attr_id set to the new "SAI_PORT_ATTR_TPID" or "SAI_LAG_ATTR_TPID"  and if the response for the corresponding new attribute_id is "implemented", it guarantee that the SAI is capable of handling at least 4 TPIDs per system?

Reply:
yes, the response will be a struct saying whether it's implemented on create, set, and get operations
and we can say it's considered implemented only if all the 4 values mentioned are supported

Reply:
Sounds good.  I am ok with this suggested change if rest of SAI community also agree.  We can then eliminate the need of this new switch attribute "SAI_SWITCH_ATTR_TPID_CONFIGURABLE"
Thanks!

Reply:
@gechiang we need to explicit lay out expectattion that if incoming packet TPID is not match Configured TPID packet should not be discarded by switch and should be classified as L2 Packet (Untag) and everything after MAC SA is considered as payload.

I also feel if above we might have to see if we need to SAI attribute to control Discard Vs Untag behaviour if it is clean approach

Reply:
I have revised the Proposal documentation to reflect what you suggested in regard to mismatched TPID handling.  Please review the new revised version.
Thanks!